src/src/com/microsoft/aad/adal/StorageHelper.java

```diff
@@ -305,70 +283,241 @@ public String decrypt(String value) throws GeneralSecurityException, IOException
-     * @throws UnsupportedEncodingException
+     * @throws IOException
      */
-    public String encrypt(String clearText)
-            throws GeneralSecurityException, IOException {
+    public synchronized SecretKey loadSecretKeyForDecryption(String keyVersion) throws GeneralSecurityException, IOException {
+        if (mSecretKeyFromAndroidKeyStore != null) {
+            return mSecretKeyFromAndroidKeyStore;
+        }
+        
+        return getKey(keyVersion);
+    }
 
-        Logger.v(TAG, "Starting encryption");
+    /**
+     * For API <18 or user provide the key, will return the user supplied key.
+     * Supported API >= 18 PrivateKey is stored in AndroidKeyStore. Loads key
+     * from the file if it exists. If not exist, it will generate one.
+     * @param keyVersion
+     * @return
+     * @throws GeneralSecurityException
+     * @throws IOException
+     */
+    private synchronized SecretKey getKeyOrCreate(final String keyVersion) throws GeneralSecurityException, IOException {
+        if (VERSION_USER_DEFINED.equals(keyVersion)) {
+            return getKey(keyVersion);
+        }
+        
+        try {
+            mSecretKeyFromAndroidKeyStore = getKey(keyVersion);
+        } catch (final IOException | GeneralSecurityException exception) {
+            // If encountering exception for reading keys, try to generate new keys
+            mKeyPair = generateKeyPairFromAndroidKeyStore();
+            
+            // Store secret key in a file after wrapping
+            File keyFile = new File(mContext.getDir(mContext.getPackageName(), Context.MODE_PRIVATE),
+                    ADALKS);
+            Cipher wrapCipher = Cipher.getInstance(WRAP_ALGORITHM);
+            // If keyfile does not exist, it needs to generate one
 
-        if (StringExtensions.IsNullOrBlank(clearText)) {
-            throw new IllegalArgumentException("Input is empty or null");
+            mSecretKeyFromAndroidKeyStore = generateSecretKey();
+            Logger.v(TAG, "Wrapping SecretKey");
+            final byte[] keyWrapped = wrap(wrapCipher, mSecretKeyFromAndroidKeyStore);
+            Logger.v(TAG, "Writing SecretKey");
+            writeKeyData(keyFile, keyWrapped);
         }
+        
+        return mSecretKeyFromAndroidKeyStore;
+    }
 
-        // load key for encryption if not loaded
-        loadSecretKeyForAPI();
-        Logger.v(TAG, "Encrypt version:" + mBlobVersion);
-        final byte[] blobVersion = mBlobVersion.getBytes(AuthenticationConstants.ENCODING_UTF8);
-        final byte[] bytes = clearText.getBytes(AuthenticationConstants.ENCODING_UTF8);
+    /**
+     * Get the saved key. Will only do read operation. 
+     * @param keyVersion
+     * @return
+     * @throws GeneralSecurityException
+     * @throws IOException
+     */
+    private synchronized SecretKey getKey(final String keyVersion) throws GeneralSecurityException, IOException {
+        if (VERSION_USER_DEFINED.equals(keyVersion)) {
```

Comment:
Same as above, the IOException will be bubbled up, clientAnalystics is added in the caller.